### PR TITLE
fix: Show println output before an error occurs in `nargo execute`

### LINKED ISSUE #3207: In `#[test]`s, stop `assert();` swallowing all `println();` output, add `assert();` template message strings and default parameter output
### Problem

I feel using `assert();` in `#[test]`s and generally testing in Noir is frustrating currently in all but the most trivial of cases, context:

- `assert();` swallows all printable output _before_ it prints meaning the only way to inspect intermediate values is to comment the `assert();` and stub the test with something like `assert(0 == 0);`. This may be fine for very trivial cases but if tests are done over arrays of data it results in a lot of noise. **Proposed fix**: `assert();` should NOT stop `println();` from outputting and should respect it's place within source files instead of being (what I assume is) optimistically executed and/or blocking said output regardless.

```rust
// ... logic
println(f"State {foo} / {bar}"); // Until a better debugging story.
assert(foo == bar); // If this triggers the above println gets swallowed. Comment this line to get it back, now do that for every single failing test every single time they fail just to be able to see the values.
```

- `assert();` can now take optional assertion messages. These are nice but I posit only useful in trivial circumstances because they only accept literal string values. This is intimately tied to the next problem.

```rust
assert(foo == bar, f"Expected equal integer values {foo} == {bar}"); // Cannot do this, it's not a string literal. Cannot `println` it as those are swallowed.
```

- `assert();` does not show the failed constraint values in tests. This is the big one. Swallowing `println();` statements would be fine, and `assert();` only accepting string literal message values would be fine if it showed the values that failed assertion. If you have a literal assert as follows: `assert(5 == 2);` then this is no issue as it is extremely obvious that 5 does not equal 2. Now imagine one of those values is a parameter or a variable `assert(foo == 2);`. Still _mostly_ fine, it's obvious that `foo` does not equal 2. What value is `foo` though? You cannot get this information. You cannot `println();` this information without commenting out the assert. You cannot message this information via `assert(foo == 2, f"Expected {foo} to equal 2");` as that's not a string literal. Now imagine you have two computed values `assert(foo == bar);`. What values are these? Impossible to tell unless you make aforementioned significant modifications to your code simply to inspect a failing test case where said modifications are mutually exclusive with having a test in the first place! **Proposed fix**: when an `assert();` fails in a `#[test]` it should print the values of it's parameters.

## Template assertion messages

This code:

```rust
let foo = 5;
let bar = 2;
assert(foo == bar, f"Expected equal integer values {foo} == {bar}");
```

Should compile and print in the existing assertion error format:

```
[big_project] Testing something_foo_bar... error: Expected equal integer values 5 == 2
     ┌─ /path/to/source/code.nr:1337:69
     │
1337 │         assert(foo == bar);
     │                ----------
```

and NOT:

```
[big_project] Testing something_foo_bar... error: Failed constraint
     ┌─ /path/to/source/code.nr:1337:69
     │
1337 │         assert(foo == bar);
     │                ----------
```

## Assertion parameter values output by default

```rust
let foo = 5;
let bar = 2;
assert(foo == bar);
```

Should print in the existing assertion error format:

```
[big_project] Testing something_foo_bar... error: Failed constraint
     ┌─ /path/to/source/code.nr:1337:69
     │
1337 │         assert(foo == bar);
     │                ----------
     = Parameter values:
       foo: 5
       bar: 2
     = Call stack:
       1. /path/to/source/code.nr:420:11
       2. /path/to/source/code.nr:1337:69
```

and NOT:

```
[big_project] Testing something_foo_bar... error: Failed constraint
     ┌─ /path/to/source/code.nr:1337:69
     │
1337 │         assert(foo == bar);
     │                ----------
```

## Assertion should NOT swallow println output

```rust
let foo = 5;
let bar = 2;
println("Blah blah");
println(f"My foo is {foo} and my bar is {bar}");
assert(foo == bar);
```

Should print a better assertion failure as already described and keep the println output:

```
Blah blah
My foo is 5 and my bar is 2
[big_project] Testing something_foo_bar... error: Failed constraint
     ┌─ /path/to/source/code.nr:1337:69
     │
1337 │         assert(foo == bar);
     │                ----------
     = Parameter values:
       foo: 5
       bar: 2
     = Call stack:
       1. /path/to/source/code.nr:420:11
       2. /path/to/source/code.nr:1337:69
```

and NOT:

```
[big_project] Testing something_foo_bar... error: Failed constraint
     ┌─ /path/to/source/code.nr:1337:69
     │
1337 │         assert(foo == bar);
     │                ----------
```

### Alternatives Considered

I augment a script I have to produce saner Noir output but it requires stubbing out `assert()` since using `assert()` destroys everything. Script here: https://github.com/tsujp/tikan/blob/master/circuit/lib/test.sh

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Maybe

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Running code that is known to fail will hit an early assert false check in acir which leads to a runtime error which will be issued before running the program and displaying any `println` output up to that point.

Resolves #3207

## Summary\*

I have removed the early error check mentioned above. The error message for the test program:

```rs
#[test]
fn main() {
    let foo = 5;
    let bar = 2;
    println("Blah blah");
    println(f"My foo is {foo} and my bar is {bar}");
    assert(foo == bar);
}
```
Now correctly shows the printlns before the failed assert error message:

```
"Blah blah"
My foo is 0x05 and my bar is 0x02
error: Failed constraint
  ┌─ /Users/../short/src/main.nr:9:12
  │
9 │     assert(foo == bar);
  │            ----------
  │
  = Call stack:
    1. /Users/../short/src/main.nr:9:12

Cannot satisfy constraint
exit 1
```
Re-running the program again however gives us a worse error message:
```rs
"Blah blah"
My foo is 0x05 and my bar is 0x02
Cannot satisfy constraint
exit 1
```
It seems after caching the program, the location for the error is lost.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by vezenovm** [comment]
The error is nicer but would it be better if we just improve the output for early assert false check?

**Comment by jfecher** [comment]
I think this particular check is just unneeded

File: compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs [hunk 1]
```diff
@@ -427,20 +427,10 @@ impl AcirContext {
...
-            } else {
-                // Constraint is always false - this program is unprovable.
-                return Err(RuntimeError::FailedConstraint {
-                    lhs: Box::new(lhs_expr),
-                    rhs: Box::new(rhs_expr),
-                    call_stack: self.get_call_stack(),
-                    assert_message,
-                });
-            };
+        if diff_expr.is_const() && diff_expr.is_zero() {
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs:430]:
```suggestion
        if diff_expr.is_zero() {
```

This implies the other condition so we can remove it. We only had the other condition in order to throw the error if the expression was non-zero.

**Comment by jfecher** [comment]
Closing this PR, after some more investigation I think the worse error message may be related to caching the results of programs after recompiling `nargo` instead.

**Comment by jfecher** [comment]
Reopening this as it is also the cause of #3207. Merging this will fix #3207 and show the output of the program up to the error. The different error format though was a separate issue that I think may be related to versioning/caching.

**Comment by jfecher** [comment]
Reopening this as it is also the cause of https://github.com/noir-lang/noir/issues/3207. Merging this will fix https://github.com/noir-lang/noir/issues/3207 and show the output of the program up to the error. The different error format though was a separate issue that I think may be related to versioning/caching.

**Comment by jfecher** [comment]
Reopening this as it is also the cause of https://github.com/noir-lang/noir/issues/3207. Merging this will fix https://github.com/noir-lang/noir/issues/3207 and show the output of the program up to the error. The different error format though was a separate issue that I think may be related to versioning/caching.

**Comment by jfecher** [comment]
@TomAFrench any insights on the caching issue?

**Comment by jfecher** [comment]
Confirmed that after merging master again, I'm still seeing the location on the error message being dropped:
```
~/noir/short> nargo execute
"Blah blah"
My foo is 0x05 and my bar is 0x02
error: Failed constraint
  ┌─ /../short/src/main.nr:9:12
  │
9 │     assert(foo == bar);
  │            ----------
  │
  = Call stack:
    1. /../noir/short/src/main.nr:9:12

Cannot satisfy constraint
exit 1

~/noir/short> nargo execute
"Blah blah"
My foo is 0x05 and my bar is 0x02
Cannot satisfy constraint
exit 1
```

**Comment by TomAFrench** [comment]
This issue is fixed by https://github.com/noir-lang/noir/pull/3133

**Comment by jfecher** [comment]
This PR is ready for review then since the error in question is unrelated

**Comment by kevaundray** [comment]
@vezenovm could you re-review?

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**